libyul/AsmAnalysis.cpp

```diff
@@ -382,191 +341,102 @@ bool AsmAnalyzer::operator()(Switch const& _switch)
-
-	m_stackHeight = initialHeight;
-
-	return success;
 }
 
-bool AsmAnalyzer::operator()(ForLoop const& _for)
+void AsmAnalyzer::operator()(ForLoop const& _for)
 {
 	yulAssert(_for.condition, "");
 
 	Scope* outerScope = m_currentScope;
 
-	int const initialHeight = m_stackHeight;
+	(*this)(_for.pre);
 
-	bool success = true;
-	if (!(*this)(_for.pre))
-		success = false;
 	// The block was closed already, but we re-open it again and stuff the
 	// condition, the body and the post part inside.
-	m_stackHeight += scope(&_for.pre).numberOfVariables();
 	m_currentScope = &scope(&_for.pre);
 
-	if (!expectExpression(*_for.condition))
-		success = false;
-
-	m_stackHeight--;
+	expectExpression(*_for.condition);
 
 	// backup outer for-loop & create new state
 	auto outerForLoop = m_currentForLoop;
 	m_currentForLoop = &_for;
 
-	if (!(*this)(_for.body))
-		success = false;
-
-	if (!(*this)(_for.post))
-		success = false;
+	(*this)(_for.body);
+	(*this)(_for.post);
 
-	m_stackHeight = initialHeight;
 	m_currentScope = outerScope;
 	m_currentForLoop = outerForLoop;
-
-	return success;
 }
 
-bool AsmAnalyzer::operator()(Block const& _block)
+void AsmAnalyzer::operator()(Block const& _block)
 {
-	bool success = true;
 	auto previousScope = m_currentScope;
 	m_currentScope = &scope(&_block);
 
-	int const initialStackHeight = m_stackHeight;
-
 	for (auto const& s: _block.statements)
-		if (!std::visit(*this, s))
-			success = false;
-
-	m_stackHeight -= scope(&_block).numberOfVariables();
-
-	int const stackDiff = m_stackHeight - initialStackHeight;
-	if (success && stackDiff != 0)
-	{
-		m_errorReporter.declarationError(
-			_block.location,
-			"Unbalanced stack at the end of a block: " +
-			(
-				stackDiff > 0 ?
-				to_string(stackDiff) + string(" surplus item(s).") :
-				to_string(-stackDiff) + string(" missing item(s).")
-			)
-		);
-		success = false;
-	}
+		std::visit(*this, s);
 
 	m_currentScope = previousScope;
-	return success;
-}
-
-bool AsmAnalyzer::expectExpression(Expression const& _expr)
-{
-	bool success = true;
-	int const initialHeight = m_stackHeight;
-	if (!std::visit(*this, _expr))
-		success = false;
-	if (success && !expectDeposit(1, initialHeight, locationOf(_expr)))
-		success = false;
-	return success;
 }
 
-bool AsmAnalyzer::expectDeposit(int _deposit, int _oldHeight, SourceLocation const& _location)
+YulString AsmAnalyzer::expectExpression(Expression const& _expr)
 {
-	if (m_stackHeight - _oldHeight != _deposit)
-	{
-		m_errorReporter.typeError(
-			_location,
-			"Expected expression to return one item to the stack, but did return " +
-			to_string(m_stackHeight - _oldHeight) +
-			" items."
+	vector<YulString> types = std::visit(*this, _expr);
+	if (types.size() != 1)
+		typeError(
+			locationOf(_expr),
+			"Expected expression to evaluate to one value, but got " +
+			to_string(types.size()) +
+			" values instead."
 		);
-		return false;
-	}
-	return true;
+	return types.empty() ? m_dialect.defaultType : types.front();
```

Comment:
Does returning only `front()` here always work?

Reply:
It is a means to at least return something. It can cause subsequent errors, but better than a fatal error, I would say.